File: binaryTreeInorderTraversal/binaryTreeInorderTraversal.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution{
public:
  vector<int> inorderTraversal(TreeNode *root) {
    vector<int> ans;
    if (!root)
      return ans;
    stack<TreeNode*> s;
    TreeNode *r = root;
    while (r || !s.empty()) {
      while (r) {
	s.push(r);
	r = r->left;
      }
      r = s.top();
      s.pop();
      ans.push_back(r->val);
      r = r->right;
    }
    return ans;
  }
};

int main()
{
  Solution s;
  int tree[12] = {1,2,3,4,9999,5,6,9999,9999,9999,9999,7};
  int size = sizeof(tree)/sizeof(tree[0]);
  vector<TreeNode *> vec;
  for (int i = 0; i < size; i++) {
    if (tree[i] != 9999) {
      vec.push_back(new TreeNode(tree[i]));
    } else {
      vec.push_back(NULL);
    }
  }
  for (int i = 0; i < size/2; i++) {
    if(!vec[i])
      continue;
    if (i*2 + 1 < size)
      vec[i]->left = vec[i*2 + 1];
    if (i*2 + 2 < size)
      vec[i]->right = vec[i*2 + 2];
  }
  vector<int> res = s.inorderTraversal(vec[0]);
  for (int i : res)
    cout << i << " ";
  cout << endl;
  for (TreeNode *t : vec) {
    //cout << t->val << " ";
    delete t;
  }
  vec.clear();
  return 0;
}

File: binaryTreeInorderTraversal/binaryTreeInorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public class Node {
        Node(TreeNode t, int op) {
            this.t = t;
            this.op = op;
        }
        TreeNode t;
        int op; // 0 -> print, 1 -> visit
    }
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<Integer>();
        if (root == null) return res;
        Stack<Node> stk = new Stack<Node>();
        stk.add(new Node(root, 1));
        while (!stk.isEmpty()) {
            Node n = stk.pop();
            if (n.op == 0) {
                res.add(n.t.val);
            } else if (n.op == 1) {
                if (n.t.right != null) stk.add(new Node(n.t.right, 1));
                stk.add(new Node(n.t, 0));
                if (n.t.left != null) stk.add(new Node(n.t.left, 1));
            }
        }
        return res;
    }
}
